LibOS/shim/src/bookkeep/shim_vma.c

```diff
@@ -37,202 +37,217 @@
 #include <errno.h>
 #include <stdbool.h>
 
-unsigned long mem_max_npages __attribute_migratable = DEFAULT_MEM_MAX_NPAGES;
-
-static void * heap_top, * heap_bottom;
+/*
+ * Internal bookkeeping for VMAs (virtual memory areas). This data
+ * structure can only be accessed in this source file, with vma_list_lock
+ * held. No reference counting needed in this data structure.
+ */
+DEFINE_LIST(shim_vma);
+struct shim_vma {
+    LIST_TYPE(shim_vma)     list;
+    void *                  start;
+    void *                  end;
+    int                     prot;
+    int                     flags;
+    uint64_t                offset;
+    struct shim_handle *    file;
+    char                    comment[VMA_COMMENT_LEN];
+};
 
 #define VMA_MGR_ALLOC   64
 #define PAGE_SIZE       allocsize
 
-static LOCKTYPE vma_mgr_lock;
-
-#define system_lock()       lock(vma_mgr_lock)
-#define system_unlock()     unlock(vma_mgr_lock)
-
-static inline void * __vma_malloc (size_t size)
+/*
+ * Because the default system_malloc() must create VMA(s), we need
+ * a new system_malloc() to avoid cicular dependency. This __malloc()
+ * stores the VMA address and size in the current thread to delay the
+ * bookkeeping until the allocator finishes extension.
```

Comment:
How confident are you that you won't need to make a second VMA on any code paths here?  In other words, is one thread entry enough, or can you add some assertion?

Reply:
I changed the design in the later commit. Now I reserve 4 VMAs, and refill them when the bookkeeping is finished. I am quite confident that we wouldn't need more than 4, but we can make this number even bigger if we want to avoid the risk.

Reply:
I would add an assert that you don't overflow, and update the comments.

Reply:
The assertion is already added.